compat.c

```diff
@@ -156,53 +156,80 @@ compat_banner(struct ssh *ssh, const char *version)
 	debug_f("no match: %s", version);
 }
 
+/* Always returns pointer to allocated memory */
 char *
 compat_cipher_proposal(struct ssh *ssh, char *cipher_prop)
 {
-	if (!(ssh->compat & SSH_BUG_BIGENDIANAES))
-		return cipher_prop;
+	char *p;
+	if (!(ssh->compat & SSH_BUG_BIGENDIANAES)) {
+		p = strdup(cipher_prop);
+		if (p == NULL)
+			fatal("memory allocation failed");
+		return p;
+	}
 	debug2_f("original cipher proposal: %s", cipher_prop);
-	if ((cipher_prop = match_filter_denylist(cipher_prop, "aes*")) == NULL)
+	p = match_filter_denylist(cipher_prop, "aes*");
+	if (p == NULL)
 		fatal("match_filter_denylist failed");
-	debug2_f("compat cipher proposal: %s", cipher_prop);
-	if (*cipher_prop == '\0')
+	debug2_f("compat cipher proposal: %s", p);
+	if (*p == '\0')
 		fatal("No supported ciphers found");
-	return cipher_prop;
+	return p;
 }
 
+/* Always returns pointer to allocated memory */
 char *
 compat_pkalg_proposal(struct ssh *ssh, char *pkalg_prop)
 {
-	if (!(ssh->compat & SSH_BUG_RSASIGMD5))
-		return pkalg_prop;
+	char *p;
+	if (!(ssh->compat & SSH_BUG_RSASIGMD5)) {
+		p = strdup(pkalg_prop);
+		if (p == NULL)
+			fatal("memory allocation failed");
+		return p;
+	}
 	debug2_f("original public key proposal: %s", pkalg_prop);
-	if ((pkalg_prop = match_filter_denylist(pkalg_prop, "ssh-rsa")) == NULL)
+	p = match_filter_denylist(pkalg_prop, "ssh-rsa");
+	if (p == NULL)
 		fatal("match_filter_denylist failed");
-	debug2_f("compat public key proposal: %s", pkalg_prop);
-	if (*pkalg_prop == '\0')
+	debug2_f("compat public key proposal: %s", p);
+	if (*p == '\0')
 		fatal("No supported PK algorithms found");
-	return pkalg_prop;
+	return p;
 }
 
+/* Always returns pointer to allocated memory */
 char *
 compat_kex_proposal(struct ssh *ssh, char *p)
```

Comment:
if you change the argument from p to prop and use p as the newly allocated string below, fewer of the lines below will need to change and the diff will be easier to read (plus it'll be more consistent with the functions above).